payjoin/src/receive/v2/session.rs

```diff
@@ -0,0 +1,512 @@
+use std::time::SystemTime;
+
+use serde::{Deserialize, Serialize};
+
+use super::{Receiver, ReceiverTypeState, SessionContext, UninitializedReceiver};
+use crate::output_substitution::OutputSubstitution;
+use crate::persist::SessionPersister;
+use crate::receive::v2::{extract_err_req, subdir, SessionError};
+use crate::receive::{v1, JsonReply};
+use crate::{ImplementationError, IntoUrl, PjUri, Request};
+
+/// Errors that can occur when replaying a receiver event log
+#[derive(Debug)]
+pub struct ReplayError(InternalReplayError);
+
+impl std::fmt::Display for ReplayError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        use InternalReplayError::*;
+        match &self.0 {
+            SessionExpired(expiry) => write!(f, "Session expired at {expiry:?}"),
+            InvalidStateAndEvent(state, event) => write!(
+                f,
+                "Invalid combination of state ({state:?}) and event ({event:?}) during replay",
+            ),
+            PersistenceFailure(e) => write!(f, "Persistence failure: {e}"),
+        }
+    }
+}
+impl std::error::Error for ReplayError {}
+
+impl From<InternalReplayError> for ReplayError {
+    fn from(e: InternalReplayError) -> Self { ReplayError(e) }
+}
+
+#[derive(Debug)]
+pub(crate) enum InternalReplayError {
+    /// Session expired
+    SessionExpired(SystemTime),
```

Comment:
Similar to Terminal State on ReceiverTypeState, should expired be a variant on the reciever type?